kore/src/Kore/Builtin/Map.hs

```diff
@@ -561,3 +581,109 @@ unifyEquals unifyEqualsChildren first second = do
+    return InKeys { symbol, keyTerm, mapTerm }
+matchInKeys _ = Nothing
+
+unifyNotInKeys
+    :: forall variable unifier
+    .  InternalVariable variable
+    => MonadUnify unifier
+    => TermSimplifier variable unifier
+    -> NotSimplifier unifier
+    -> CeilSimplifier unifier (TermLike variable) (OrCondition variable)
+    -> TermLike variable
+    -> TermLike variable
+    -> MaybeT unifier (Pattern variable)
+unifyNotInKeys
+    unifyChildren
+    (NotSimplifier notSimplifier)
+    (CeilSimplifier ceilSimplifier)
+    a
+    b
+  =
+    worker a b <|> worker b a
+  where
+    normalizedOrBottom
+       :: TermLike variable
+       -> Ac.NormalizedOrBottom Domain.NormalizedMap variable
+    normalizedOrBottom = Ac.toNormalized
+
+    defineTerm :: TermLike variable -> MaybeT unifier (Condition variable)
+    defineTerm termLike =
+        ceilSimplifier
+            TermLike.Ceil
+                { ceilResultSort = Sort.predicateSort
+                , ceilOperandSort = TermLike.termLikeSort termLike
+                , ceilChild = termLike
+                }
+        >>= lift . Unify.scatter
+
+    eraseTerm =
+        Pattern.fromCondition . Pattern.withoutTerm
+
+    unifyAndNegate t1 t2 = do
+        unificationSolutions <-
+            fmap eraseTerm
+            <$> Unify.gather (unifyChildren t1 t2)
+        notSimplifier
+            SideCondition.top
+            (OrPattern.fromPatterns unificationSolutions)
+        >>= Unify.scatter
+
+    collectConditions terms =
+        foldr
+            (flip Pattern.andCondition)
+            Pattern.top
+            terms
+
+    worker
+        :: TermLike variable
+        -> TermLike variable
+        -> MaybeT unifier (Pattern variable)
+    worker termLike1 termLike2
+      | Just boolValue <- Bool.matchBool termLike1
+      , not boolValue
+      , Just InKeys { symbol, keyTerm, mapTerm } <- matchInKeys termLike2
+      , Ac.Normalized normalizedMap <- normalizedOrBottom mapTerm
+      = do
+        let symbolicKeys = Domain.getSymbolicKeysOfAc normalizedMap
+            concreteKeys =
+                TermLike.fromConcrete
+                <$> Domain.getConcreteKeysOfAc normalizedMap
+            mapKeys = symbolicKeys <> concreteKeys
+
+            mapValues =
+                Domain.getMapValue
+                <$> Domain.getSymbolicValuesOfAc normalizedMap
+                <> Domain.getConcreteValuesOfAc normalizedMap
+
+            opaqueElements = Domain.opaque . Domain.unwrapAc $ normalizedMap
+            keyInKeysOpaque =
+                (\term -> TermLike.mkApplySymbol symbol [keyTerm, term])
+                <$> opaqueElements
+
+        Monad.guard . not . null $ mapKeys
+        -- Concrete keys are constructor-like, therefore they are defined
+        TermLike.assertConstructorLikeKeys concreteKeys $ return ()
+        definedKeys <- traverse defineTerm (keyTerm : symbolicKeys)
+        definedValues <- traverse defineTerm mapValues
```

Comment:
Instead of checking that the individual keys and values are defined, we should evaluate `\ceil(termLike2)`. This will recursively call the `\ceil` rule defined in the frontend.